apps/meteor/app/lib/server/functions/saveUser/saveUser.ts

```diff
@@ -44,135 +47,148 @@ export type SaveUserData = {
-	}
-
-	if (userData.language) {
-		updater.set('language', userData.language);
-	}
-
-	if (typeof userData.requirePasswordChange !== 'undefined') {
-		updater.set('requirePasswordChange', userData.requirePasswordChange);
-		if (!userData.requirePasswordChange) {
-			updater.unset('requirePasswordChangeReason');
+
+		handleBio(updater, userData.bio);
+		handleNickname(updater, userData.nickname);
+
+		if (userData.roles) {
+			updater.set('roles', userData.roles);
 		}
-	}
-
-	if (typeof userData.verified === 'boolean' && !userData.email) {
-		updater.set('emails.0.verified', userData.verified);
-	}
-
-	await Users.updateFromUpdater({ _id: userData._id }, updater);
-
-	// App IPostUserUpdated event hook
-	const userUpdated = await Users.findOneById(userData._id);
-
-	await callbacks.run('afterSaveUser', {
-		user: userUpdated,
-		oldUser: oldUserData,
-	});
-
-	await Apps.self?.triggerEvent(AppEvents.IPostUserUpdated, {
-		user: userUpdated,
-		previousUser: oldUserData,
-		performedBy: await safeGetMeteorUser(),
-	});
-
-	if (sendPassword) {
-		await sendPasswordEmail(userData);
-	}
-
-	if (typeof userData.verified === 'boolean') {
-		delete userData.verified;
-	}
-	void notifyOnUserChange({
-		clientAction: 'updated',
-		id: userData._id,
-		diff: {
-			...userData,
-			emails: userUpdated?.emails,
-		},
-	});
-
-	return true;
-};
+		if (userData.settings) {
+			updater.set('settings', { preferences: userData.settings.preferences });
+		}
+
+		if (userData.language) {
+			updater.set('language', userData.language);
+		}
+
+		if (typeof userData.requirePasswordChange !== 'undefined') {
+			updater.set('requirePasswordChange', userData.requirePasswordChange);
+			if (!userData.requirePasswordChange) {
+				updater.unset('requirePasswordChangeReason');
+			}
+		}
+
+		if (typeof userData.verified === 'boolean' && !userData.email) {
+			updater.set('emails.0.verified', userData.verified);
+		}
+
+		if (userData.customFields) {
+			await saveCustomFields(userData._id, userData.customFields, { _updater: updater, session });
+		}
+
+		await Users.updateFromUpdater({ _id: userData._id }, updater, { session });
+
+		// App IPostUserUpdated event hook
+		// We need to pass the session here to ensure this record is fetched
+		// with the uncommited transaction data.
+		const userUpdated = await Users.findOneById(userData._id, { session });
+
+		await callbacks.run('afterSaveUser', {
+			user: userUpdated,
+			oldUser: oldUserData,
+		});
+
+		await Apps.self?.triggerEvent(AppEvents.IPostUserUpdated, {
+			user: userUpdated,
+			previousUser: oldUserData,
+			performedBy: await safeGetMeteorUser(),
+		});
+
+		if (sendPassword) {
+			await sendPasswordEmail(userData);
+		}
+
+		if (typeof userData.verified === 'boolean') {
+			delete userData.verified;
+		}
+		void notifyOnUserChange({
+			clientAction: 'updated',
+			id: userData._id,
+			diff: {
+				...userData,
+				emails: userUpdated?.emails,
+			},
+		});
```

Comment:
I wonder if we could move these to happen after the transaction has commited.
For example: if the transaction aborts outside of this, we'll call the sendPasswordEmail again when the transaction happens again.

The tx should contain the minimal operations required to keep state, but not tell others about that state until after commit.

Reply:
I agree, but this would require much more effort. It's a concern in other functions aswell. This function has a very deep call stack, and a bunch of those methods have side effects that are not so easy to externalize. I'll think about this some more